### ISSUE TITLE
FEATURE: Tools for race condition prevention

### ISSUE THREAD

**Issue by RasmusSkytte** [ssi-dk/SCDB#63]
**Describe the feature**

In short, a data base version of [flock](https://cran.r-project.org/web/packages/flock/index.html).
In `update_snapshot` several data base operations occur in sequence: `rows_update`, `rows_append`, `rows_delete` etc.
In some cases, parallel calls to `update_snapshot` could result in race conditions between these calls.

For `diseasystore` I would like to use such functions and have developed a first draft of such a system.
https://github.com/ssi-dk/diseasystore/issues/74

Should we add these functions to `SCDB` instead?
Should we make a small package `dblock` that just has these functions?

**What use will this feature have?**
See https://github.com/ssi-dk/diseasystore/blob/database_locks/R/0_locks.R for the first draft.

In short, the feature consist of four functions
`add_table_lock()` a function that assigns the current process id (pid) to a given table
`remove_table_lock()` a function that removes the current process id (pid) from a given table
`get_lock_owner()` a function that returns the current process id (pid) assigned to a given table
`cleanup_locks()` a function that removes old locks.

**Who should add the feature?**
I could do it after some discussion of the draft solution

**Additional context**
N/A

**Comment by marcusmunch** [comment]
It makes a lot of sense to keep this within the scope of SCDB or its dependencies. The root cause IMO is `update_snapshot` not running the individual steps in any sort of transaction block, and I therefore think it would make sense to resolve this within SCDB.

I would resolve the issue as follows (using myschema.mytable as example):

1. Create a table `myschema.mytable_lock`.
2. Write a call to `on.exit` within `update_snapshot` which drops the table.
3. Before trying any `INSERT` or `UPDATE` operations, check if `myschema.mytable_lock` exists. If it does, raise an error.

This approach does require CREATE TABLE privileges within the database, which we should potentially be careful to presume.

**Comment by marcusmunch** [comment]
BONUS ROUND: Turns out `DBI::dbWithTransaction` is a thing.

Enjoy 😄
